### DISCUSSION TITLE
WooCommerce FTS Indexes for HPOS

### DISCUSSION THREAD

**Discussion by vedanshujain** [woocommerce/woocommerce#48138] [WooCommerce HPOS]
We are introducing FTS indexes with 9.0 in experimental mode. Let's chat about issues, comments, and any other feedback about this feature!

**Comment by ignacyhologa** [comment]
Hi I enabled the FTS indexes on my site and there is a feature missing in the search. Before I was able to search by 4 last digits of the order number and it found results. Now it is not finding any order. 

To be clear, the order number looks sth like this: F0248375 and it is not the order id. 

It appears to work only when i select the category of the search and select 'order number' there but this adds another step and is thus slower. Since the default selection is 'All' it should work in the default, right? 

Additionally, it would be great to have a filter hook allowing to set the default field in the selector fo the type of the search as we mostly use it in one way. 

Another request would be to add a country selector so i can filter the results down by country.

**Reply by vedanshujain** [reply]
Thanks for the response, do you know where is the order number stored? Is it part of the metadata?

**Reply by vedanshujain** [reply]
We don't have a dedicated field of order number, so I am thinking it might be getting added by a plugin perhaps. 

> it would be great to have a filter hook allowing to set the default field in the selector fo the type of search as we mostly use it in one way.

It should remember the field you last used (from 8.9.0 onwards) and it saves that data in the session cookie. But also happy to add a filter to customize the default value.

> add a country selector so i can filter the results down by country.

I think we can explore a very rich search as a future enhancement, thanks for the suggestion!

**Reply by ignacyhologa** [reply]
Thank for the response. It is indeed added by a plugin called https://wordpress.org/plugins/wt-woocommerce-sequential-order-numbers. I am not sure how they store it exactly. Anyways, it would be great to integrate this if possible, else just adding a filter as discussed for the search type would be great.

> It should remember the field you last used (from 8.9.0 onwards) and it saves that data in the session cookie. But also happy to add afilter to customize the default value.

I checked the field last used as described by you, but it seems to not work for my field 'order number'. It works for the other fields. So when i filter by the 'order number' and then load the orders again it switches to the 'order id' in the filter selector. 

> I think we can explore a very rich search as a future enhancement, thanks for the suggestion!
That would be awesome addition and help for shops that sell to many countries. For example, we often manage orders destined for the same country and process them together and this filter would allow for quick access to these.

**Comment by dihan29** [comment]
Hello, The order number search works with hpos enabled but doesn’t works with new FTS in Woocommerce mobile app or website. It shows order id only. Kindly make it more useable with better integration for sequential order number plugin. It works with normal hpos enabled and without new search index enabled. It shows order id only instead of order number when enabled. Kindly make it full support for sequential order number plugin with latest index for better performance. It uses “order_number” in code instead of “order_id”

$order_number = $order->get_order_number

![Screenshot_2024-06-23-22-08-23-678_com woocommerce android](https://github.com/woocommerce/woocommerce/assets/173668451/6a1189f7-2aff-4ec2-8656-91ef19d49343)

**Reply by vedanshujain** [reply]
Thanks, let me checkout this plugin to see how can we add the support (here but also in general for other plugins that extend the search functionality).

**Reply by dihan29** [reply]
Thank you. Hoping for quick fix as it works already with HPOS enabled without FTS. I have also informed sequential plugin creator on wocommerce forum. They have compatibility till 8.9.3 they are also trying to fix. Thanks
https://wordpress.org/support/topic/order-number-search-doesnt-work-on-wocommerce-mobile/#post-17848036

**Reply by vedanshujain** [reply]
I see the issue here, basically the search integration should be done using the combination of `woocommerce_hpos_admin_search_filters` and `woocommerce_hpos_generate_where_for_search_filter` instead of the JS and meta query overwriting like it's currently being done in the order number search plugin. We will add a doc for it!

**Reply by dihan29** [reply]
Any update or fix brother?

**Reply by vedanshujain** [reply]
Note that the upstream plugin has published an update that should fix it.

**Reply by dihan29** [reply]
Still doesn't work with the update. FTS search doesn't show proper order number. Using with disabled FTS for now.

**Comment by vanloc0301** [comment]
After enabling this feature in Woo, I noticed that searching for orders on the Order page became very slow. 
![image](https://github.com/woocommerce/woocommerce/assets/4168981/5485eb38-8e08-41af-b572-e95ad70ac298)
I’ve checked multiple times, and when I turn off WooCommerce FTS Indexes for HPOS, the search is slower but still within a range of 4-7 seconds. However, if I leave this feature enabled, the search time increases significantly to a range of 21-25 seconds

**Reply by vanloc0301** [reply]
Edited: this only slow when I search keyword like: "xxxx" with xxxx is name of Order (example in my site have many order relate with product have name: `Adobe`) I want to show list all order relate with Adobe product.
Opposite, when I try search with Order ID like: `14381`. It very fast. Range 0.5 - 2.5 seconds.
Seem the problem only affect when I search with keyword, not affect if I search with number.
I using monitor capture SQL query and check all query, don't have any query have keyword: `Adobe` in result of report. Only show query when I search Order ID: `14381`, you can see this image:
![image](https://github.com/woocommerce/woocommerce/assets/4168981/d1db7cb4-cb89-4411-9418-14f835617896)

**Reply by vedanshujain** [reply]
This is interesting, just to confirm, so the query you have mentioned in the screenshot is fast, but when you replace it with something like Adobe you are seeing much slower results.

Can I ask you to share your MySQL version and output of the explain query? If you have access to run the query directly, you can run the explain query by adding a `EXPLAIN` keyword, like `EXPLAIN SELECT ...`.

**Reply by vanloc0301** [reply]
Yes. When searching for an `Order-ID`, the results load quickly. 
However, when I search for keywords like ‘Adobe’ or anything related to ‘Stock’ or ‘Flower stock,’ it takes much longer to display the results. 

I’m currently using Woocommerce version 9.0.2 and MariaDB version 11.4.2, which I recently upgraded.

Explain query result for query:
```
SELECT xkk_wc_orders.id FROM xkk_wc_orders  WHERE 1=1 AND (xkk_wc_orders.status IN ('wc-pending','wc-processing','wc-on-hold','wc-completed','wc-cancelled','wc-refunded','wc-failed')) AND (xkk_wc_orders.type = 'shop_order') AND ( ( `xkk_wc_orders`.id = 14381 OR `xkk_wc_orders`.id = 14381 OR `xkk_wc_orders`.transaction_id LIKE '%14381%' OR `xkk_wc_orders`.billing_email LIKE '14381%' OR  xkk_wc_orders.id IN ( SELECT order_id FROM xkk_wc_order_addresses WHERE MATCH( xkk_wc_order_addresses.first_name, xkk_wc_order_addresses.last_name, xkk_wc_order_addresses.company, xkk_wc_order_addresses.address_1, xkk_wc_order_addresses.address_2, xkk_wc_order_addresses.city, xkk_wc_order_addresses.state, xkk_wc_order_addresses.postcode, xkk_wc_order_addresses.country, xkk_wc_order_addresses.email ) AGAINST ( '*14381*' IN BOOLEAN MODE ) )  OR  xkk_wc_orders.id in ( SELECT order_id FROM xkk_woocommerce_order_items search_query_items WHERE MATCH ( search_query_items.order_item_name ) AGAINST ( '*14381*' IN BOOLEAN MODE ) )  ) ) GROUP BY xkk_wc_orders.id ORDER BY xkk_wc_orders.date_created_gmt DESC LIMIT 0, 100
```
![image](https://github.com/woocommerce/woocommerce/assets/4168981/b234a4d8-0a31-4079-b339-1d321141db3b)

**Reply by vedanshujain** [reply]
It is looking at just one row, so I wonder where the slowness is coming from.

**Reply by vedanshujain** [reply]
Oh sorry, actually, I meant the explain results of the query which is slow. So with `Adobe` or `Stock` etc and not with `14381`.

**Reply by vanloc0301** [reply]
I see two query relative with "Adobe" keyword:
**### First query:**
You can see this query take long time: 3,9608ms 
```
SELECT xkk_wc_orders.id
FROM xkk_wc_orders
WHERE 1=1
AND (xkk_wc_orders.status IN ('wc-pending','wc-processing','wc-on-hold','wc-completed','wc-cancelled','wc-refunded','wc-failed'))
AND (xkk_wc_orders.type = 'shop_order')
AND ( ( `xkk_wc_orders`.transaction_id LIKE '%adobe%'
OR `xkk_wc_orders`.billing_email LIKE 'adobe%'
OR xkk_wc_orders.id IN (
SELECT order_id
FROM xkk_wc_order_addresses
WHERE MATCH( xkk_wc_order_addresses.first_name, xkk_wc_order_addresses.last_name, xkk_wc_order_addresses.company, xkk_wc_order_addresses.address_1, xkk_wc_order_addresses.address_2, xkk_wc_order_addresses.city, xkk_wc_order_addresses.state, xkk_wc_order_addresses.postcode, xkk_wc_order_addresses.country, xkk_wc_order_addresses.email ) AGAINST ( '*adobe*' IN BOOLEAN MODE ) )
OR xkk_wc_orders.id in (
SELECT order_id
FROM xkk_woocommerce_order_items search_query_items
WHERE MATCH ( search_query_items.order_item_name ) AGAINST ( '*adobe*' IN BOOLEAN MODE ) ) ) )
GROUP BY xkk_wc_orders.id
ORDER BY xkk_wc_orders.date_created_gmt DESC
LIMIT 0, 100
```
![image](https://github.com/woocommerce/woocommerce/assets/4168981/aaf5812b-bfe0-491a-835a-07745da6e41e)
**`EXPLAIN SELECT`  first query:**
![image](https://github.com/woocommerce/woocommerce/assets/4168981/5634a4f1-5009-4524-9fb1-5a4d828190cc)
========================
**### Second query:** 
You can see this query get many times to execute: 22,7587ms
```
SELECT COUNT(DISTINCT xkk_wc_orders.id)
FROM xkk_wc_orders
WHERE 1=1
AND (xkk_wc_orders.status IN ('wc-pending','wc-processing','wc-on-hold','wc-completed','wc-cancelled','wc-refunded','wc-failed'))
AND (xkk_wc_orders.type = 'shop_order')
AND ( ( `xkk_wc_orders`.transaction_id LIKE '%adobe%'
OR `xkk_wc_orders`.billing_email LIKE 'adobe%'
OR xkk_wc_orders.id IN (
SELECT order_id
FROM xkk_wc_order_addresses
WHERE MATCH( xkk_wc_order_addresses.first_name, xkk_wc_order_addresses.last_name, xkk_wc_order_addresses.company, xkk_wc_order_addresses.address_1, xkk_wc_order_addresses.address_2, xkk_wc_order_addresses.city, xkk_wc_order_addresses.state, xkk_wc_order_addresses.postcode, xkk_wc_order_addresses.country, xkk_wc_order_addresses.email ) AGAINST ( '*adobe*' IN BOOLEAN MODE ) )
OR xkk_wc_orders.id in (
SELECT order_id
FROM xkk_woocommerce_order_items search_query_items
WHERE MATCH ( search_query_items.order_item_name ) AGAINST ( '*adobe*' IN BOOLEAN MODE ) ) ) )
```
![image](https://github.com/woocommerce/woocommerce/assets/4168981/def50d1c-4955-4992-86de-a6ac53fc110d)
**`EXPLAIN SELECT`  second query:**
![image](https://github.com/woocommerce/woocommerce/assets/4168981/273986be-c554-4645-a5c1-7c6d217420a0)

**Reply by vanloc0301** [reply]
I don’t know the exact details, but I searched for the `Adobe` keyword only once, and it resulted in two queries with a time range of 21 to 25 seconds.

**Reply by vedanshujain** [reply]
@vanloc0301 I want to check if the join is what's causing the slowness, can you run this query for me please and let me know how much time it took:
```
SELECT COUNT(DISTINCT xkk_wc_orders.id)
FROM xkk_wc_orders
WHERE 1 = 1 AND
	  xkk_wc_orders.id IN (SELECT order_id
						   FROM xkk_wc_order_addresses
						   WHERE MATCH(xkk_wc_order_addresses.first_name, xkk_wc_order_addresses.last_name,
									   xkk_wc_order_addresses.company, xkk_wc_order_addresses.address_1,
									   xkk_wc_order_addresses.address_2, xkk_wc_order_addresses.city,
									   xkk_wc_order_addresses.state, xkk_wc_order_addresses.postcode,
									   xkk_wc_order_addresses.country, xkk_wc_order_addresses.email)
									   AGAINST('*adobe*' IN BOOLEAN MODE))
	OR xkk_wc_orders.id in (SELECT order_id
						   FROM xkk_woocommerce_order_items search_query_items
						   WHERE MATCH(search_query_items.order_item_name) AGAINST('*adobe*' IN BOOLEAN MODE));
```

**Reply by vanloc0301** [reply]
This query takes a long time to complete. Although I’m unsure how to check the query execution time directly, I ran this query twice and used a stopwatch to measure the exact time, which turned out to be 31 seconds to complete this query.

**Reply by vedanshujain** [reply]
This is good enough, thanks! let me investigate this a bit more deeply and get back!

**Reply by vanloc0301** [reply]
Thanks @vedanshujain . I have a question: 
Is this due to a problem with my website or can the HPOS FTPS Indexes be improved?
I'm afraid there's a conflict from the plugin's way of making queries incorrect.

**Reply by vedanshujain** [reply]
I was not able to repro this when I looked at at that time, let me pick this up again and see what can we do. 

> I'm afraid there's a conflict from the plugin's way of making queries incorrect.

What do you mean by this? Do you think that the queries are different than what they are supposed to be?

**Comment by vladsucitu** [comment]
Very nice and faster than before. Very good option. 

However, we are unable to find an order by searching by phone number. It's something we can do, or will it take some time to integrate this option as well?

If this feature is not checked, we can find an order by searching by phone number, but if it is active, the search returns no results.

**Reply by rwkyyy** [reply]
+1 same issue from current testing; but the speed is great tho!

**Reply by vladsucitu** [reply]
@vedanshujain any updates here? 
Have you been able to reproduce this behavior? Is it complicated to add the phone number to the search and return results?

**Reply by vedanshujain** [reply]
I can repro, hoping to have a fix ready pretty soon!

**Reply by rwkyyy** [reply]
@vedanshujain any ETA? search speed in HPOS is a real pain for big websites 🥹

**Reply by vedanshujain** [reply]
We have a PR merged for phone numbers, but it will be available only in the next release -> https://github.com/woocommerce/woocommerce/pull/51065. It basically involves dropping and [recreating the index](https://github.com/woocommerce/woocommerce/pull/51065/files#diff-fa7fe8c90c0e20b0f3b18b8092278ec738e4ef5d43fe8d8bfdc3f52a124aaa59R360), and then the search query will also change to [accommodate the phone number](https://github.com/woocommerce/woocommerce/pull/51065/files#diff-b4db1e20946cdfc331ebb65b633fc7717db0ec7c97a752768c00fd6e65a088fdR293).

**Reply by rwkyyy** [reply]
Thank you very much for the fast reply!

I plan to use this on a big store (~300k orders atm), are there any filters to change the data that is added in the indexes? I am mostly interested in just: order number, first + last name, phone, state and city.

**Comment by berkmh** [comment]
I just migrated over to HPOS and full text yesterday on my large (500k order) store and finding on order searches im only able to find using order number and not name or email?

**Reply by rwkyyy** [reply]
disable the FTS indexes, it's not a priority for core team apparently...
use this:
https://wordpress.org/plugins/fast-woo-order-lookup/

**Reply by vladsucitu** [reply]
I think it will be available with version 9.4 - on 11.11 (was delayed) (https://developer.woocommerce.com/changelog/)

**Reply by vedanshujain** [reply]
@berkmh there's an update to FTS indexes in 9.4 as @vladsucitu points out, but even the current implementation should be able to search for name/email. would you be able to share more details (after redacting any PII for privacy) by perhaps sharing exact phrases you are searching for, or the exact query being run. Maybe there are cases that aren't yet handled.

**Reply by berkmh** [reply]
some more observations...noticing that after I search for an order with the order number and its found, then I can locate the order by email or other. Seeing the same behavior on the front where the users are unable to see their orders until I locate within admin by the order number....not sure whats happening. the docs do state it's not compatible with mysql 8+ and im on 8.0.36 @vedanshujain @vladolaru @rwkyyy have deactivated for now and trying @rwkyyy solution

**Comment by eddr** [comment]
Hello and thanks!

The change is great for the speed, however when searching specific customer full name ( first name + more ) in the orders list page, we get all results which contain one of the phrase/term parts.

For example, when searching for "Eduardo Sánchez"
We get all orders with "Eduardo" or "Sánchez".

This is different from before. What should we do to get only orders with the phrase we requested?

Thanks

**Reply by betayazilim** [reply]
I have exactly same problem.
Is there any tricks to find exact match?

**Comment by agibson-godaddy** [comment]
Hi :)

I work on the SkyVerge Sequential Order Numbers Pro plugin. Searching by our custom order ID no longer works. Our custom order ID is stored in order meta. I've seen the above comments -- particularly this one https://github.com/woocommerce/woocommerce/discussions/48138#discussioncomment-9879155 I just wanted to clarify before changing our implementation...

For non-FTS, we're using the `woocommerce_order_table_search_query_meta_keys` filter to add our meta key to the array:

```php
public function __construct() {
    add_filter( 'woocommerce_order_table_search_query_meta_keys', [ $this, 'custom_search_fields' ] );
}

public function custom_search_fields( $search_fields ) {

	array_push( $search_fields, '_order_number_formatted' );

	return $search_fields;
}
```

This works without FTS.

But it doesn't work in FTS because `generate_where_for_meta_table()` never gets called in that context.

---

Can you confirm if there are any plans to make this existing `woocommerce_order_table_search_query_meta_keys` filter work with FTS? Or should we update to use `woocommerce_hpos_generate_where_for_search_filter` instead?

**Reply by agibson-godaddy** [reply]
We're experimenting with `woocommerce_hpos_generate_where_for_search_filter` along with a custom filter. We can get it working when the filter is explicitly selected to our custom one, but if "All" is selected, then a custom filter is never called. That's due to using this hard-coded list of core filters here: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php#L56-L65

The impact is that using "All" doesn't actually search all filters; it only searches the core ones. Curious what the reason is behind this? Would be nice if we could integrate with "All" as well.

**Comment by MajorChump** [comment]
When testing the feature it seems it differs heavily to the previous implementation. Searching for postcodes returns matches for both tokens. 

For example WA9 3AT would match with everything that is WA9 and 3AT. This can be corrected by using phrase matching in mysql, something like:

```
$against = '"' . $wpdb->esc_like( $search_term ) . '"';
$mode    = 'IN BOOLEAN MODE';

$where .= " AND MATCH (col1, col2, ...) AGAINST ( '$against' $mode ) ";
```

This would then require the search to match with all the tokens. Is there a reason its not been implemented this way?

Further research seems we're intentionally allowing partial matches in 

```
public function sanitise_boolean_fts_search_term( string $param ): string {
		// Remove any operator to prevent incorrect query and fatals, such as search starting with `++`. We can allow this in the future if we have proper validation for FTS search operators.
		// Space is allowed to provide multiple words.
		$sanitized_param = preg_replace( '/[^\p{L}\p{N}_]+/u', ' ', $param );
		if ( $sanitized_param !== $param ) {
			$param = str_replace( '"', '', $param );
			return '"' . $param . '"';
		}
		// Split the search phrase into words so that we can add operators when needed.
		$words           = explode( ' ', $param );
		$sanitized_words = array();
		foreach ( $words as $word ) {
			// Add `*` as suffix to every term so that partial matches happens.
			$word              = $word . '+';
			$sanitized_words[] = $word;
		}
		return implode( ' ', $sanitized_words );
	}
```

Whats the reason for that?
